exchanges/orderbook/incremental_updates.go

```diff
@@ -0,0 +1,234 @@
+			return d.invalidate(fmt.Errorf("%s %s %s %w: expected '%d', got '%d'", d.exchange, d.pair, d.asset, errChecksumMismatch, u.ExpectedChecksum, checksum))
+		}
+	} else if d.verifyOrderbook {
+		if err := verify(d.snapshot()); err != nil {
+			return d.invalidate(err)
+		}
+	}
+
+	return nil
+}
+
+// // TODO: Confirm no alloc as it shouldn't escape
+func (d *Depth) snapshot() *Base {
+	return &Base{
+		Bids:                   d.bidTranches.Tranches,
+		Asks:                   d.askTranches.Tranches,
+		Exchange:               d.options.exchange,
+		Pair:                   d.pair,
+		Asset:                  d.asset,
+		IsFundingRate:          d.options.isFundingRate,
+		PriceDuplication:       d.options.priceDuplication,
+		IDAlignment:            d.options.idAligned,
+		ChecksumStringRequired: d.options.checksumStringRequired,
+	}
+}
+
+// updateByIDAndAction will receive an action to execute against the orderbook it will then match by IDs instead of
+// price to perform the action
+func (d *Depth) updateByIDAndAction(u *Update) error {
+	switch u.Action {
+	case Amend:
+		if err := d.updateBidAskByID(u); err != nil {
+			return fmt.Errorf("%w %w", errAmendFailure, err)
+		}
+	case Delete:
+		// edge case for Bitfinex as their streaming endpoint duplicates deletes
+		bypassErr := d.options.exchange == "Bitfinex" && d.options.isFundingRate // TODO: Confirm this is still correct
+		if err := d.deleteBidAskByID(u, bypassErr); err != nil {
+			return fmt.Errorf("%w %w", errDeleteFailure, err)
+		}
+	case Insert:
+		if err := d.insertBidAskByID(u); err != nil {
+			return fmt.Errorf("%w %w", errInsertFailure, err)
+		}
+	case UpdateInsert:
+		if err := d.updateInsertByID(u); err != nil {
+			return fmt.Errorf("%w %w", errUpdateInsertFailure, err)
+		}
+	default:
+		return fmt.Errorf("%w [%d]", errInvalidAction, u.Action)
+	}
+	return nil
+}
+
+// updateBidAskByPrice updates the bid and ask spread by supplied updates, this will trim total length of depth level to
+// a specified supplied number
+func (d *Depth) updateBidAskByPrice(update *Update) error {
+	if update.UpdateTime.IsZero() {
+		return fmt.Errorf("%s %s %s %w", d.exchange, d.pair, d.asset, ErrLastUpdatedNotSet)
+	}
+	d.bidTranches.updateInsertByPrice(update.Bids, d.options.maxDepth)
+	d.askTranches.updateInsertByPrice(update.Asks, d.options.maxDepth)
+	d.updateAndAlert(update)
+	return nil
+}
+
+// updateBidAskByID amends details by ID
+func (d *Depth) updateBidAskByID(update *Update) error {
+	if update.UpdateTime.IsZero() {
+		return fmt.Errorf("%s %s %s %w", d.exchange, d.pair, d.asset, ErrLastUpdatedNotSet)
+	}
+	if err := d.bidTranches.updateByID(update.Bids); err != nil {
+		return err
+	}
+	if err := d.askTranches.updateByID(update.Asks); err != nil {
+		return err
+	}
+	d.updateAndAlert(update)
+	return nil
+}
+
+// deleteBidAskByID deletes a price level by ID
+func (d *Depth) deleteBidAskByID(update *Update, bypassErr bool) error {
```

Comment:
🐒 
🌴 If accepted, Also applies to the following 2 funcs
```suggestion
// delete deletes a price level by ID
func (d *Depth) delete(update *Update, bypassErr bool) error {
```

Reply:
[done](https://github.com/thrasher-corp/gocryptotrader/pull/1908/commits/f13453acb65af78f67e71aca5baae8b94e747631)